lib/puppet/functions/vault_lookup/lookup.rb

```diff
@@ -18,15 +18,28 @@ def lookup(path, vault_url = nil)
     # host is defined.
     raise Puppet::Error, "Unable to parse a hostname from #{vault_url}" unless uri.hostname
 
-    use_ssl = uri.scheme == 'https'
-    connection = Puppet::Network::HttpPool.http_instance(uri.host, uri.port, use_ssl)
+    if defined? Puppet.runtime
+      # modern Puppet HTTP client. This allows us to use the system store on >= 7.16.0
+      connection = Puppet.runtime[:http]
+      token = get_auth_token(connection, vault_url)
+      secret_response = connection.get(URI("#{vault_url}/v1/#{path}"), headers: { 'X-Vault-Token' => token }, options: { include_system_store: true })
```

Comment:
`Puppet.runtime[:http]` is a `Puppet::HTTP::Client` implementation, so it's a little confusion to assign it to `connection`. Might want to get a connection directly:

```ruby
client = Puppet.runtime[:http]
connection = client.connect(uri, options: { include_system_store: true })
```

and then pass that connection through to `get_auth_token` method. Then `connection.get` just needs the request path (not the full URL) and you don't need to pass `include_system_store` again:

```ruby
response = connection.get("/v1/#{path}", 'X-Vault-Token' => token)
```

Reply:
sorry i was on vacation. i will test this.

Reply:
@joshcooper this did not work for me. i am seeing error `Error: Failed to apply catalog: undefined method `get' for nil:NilClass`

```
client = Puppet.runtime[:http]
connection = client.connect(uri, options: { include_system_store: true })
: 
: 
secret_response = if vault_namespace.nil? || vault_namespace == ''
                        connection.get("/v1/#{path}", 'X-Vault-Token' => vault_token)
                      else
                        connection.get("/v1/#{path}", headers: { 'X-Vault-Token' => vault_token, 'X-Vault-Namespace' => vault_namespace })
                      end
```

Reply:
sorry I misspoke. `Puppet::HTTP::Client#connect` takes a block and auto-closes the connection, it doesn't return the connection to the caller. In that case, I'd just stick with the old way and call:

```ruby
client = Puppet.runtime[:http]
client.get(uri, options: { include_system_store: true })
```

Reply:
@joshcooper  @root-expert when changed to `options: { include_system_store: true }` seeing the following error. works fine and secrets are accessible when used as `options: { 'include_system_store' => true }`

error message:
```
Error: Failed to apply catalog: Received 400 response code from vault at https://vault.host:8200 for authentication (api errors: ["1 error occurred:\n\t* error running lookahead function for mfa: no client certificate found\n\n"])
```

Reply:
nope my puppet CA is completely different from the certfiicate used by Vault server

Reply:
Ok sounds good. Since the agent is able to connect and verify the vault server's cert, where is the vault server's CA cert on the agent? Was it added to `/opt/puppetlabs/puppet/ssl/certs.pem`?

Also which `puppet --version` are you using? Was it installed using the `puppet-agent` package or from a gem?

Reply:
I have been trying to use this fork as well. It stops at

```
Debug: HTTP POST https: //vault-example.net:8200/v1/auth/cert/login returned 400 Bad Request
Debug: Caching connection for https://vault-test.example.net: 8200
Error: Failed to apply catalog: Received 400 response code from vault at for authentication (api errors: ["client
certificate must be supplied"])
```

I think the ssl_context needs to be passed to provide the puppet client cert and key

Reply:
The client cert error is due to https://tickets.puppetlabs.com/browse/PUP-11522. When calling `client.get(uri, options: { include_system_store: true })` puppet does not add the client cert and private key to the generated context. Normally that's not needed when talking to third party HTTPS servers. Once that ticket is resolved, it should work when using `include_system_store` as a symbol.

Reply:
@tw3rp @vamsinm can you provide some details about how you have vault configured? I can be reached at @joshcooper at https://puppet.slack.com in the #puppet-dev channel or DM if that's better for you.

Reply:
@joshcooper I need an invitation to join puppet.slack.com

Reply:
How would the client get the vault token if the client cert and private key are not passed on the the login post method.

Reply:
Sorry I meant https://puppetcommunity.slack.com/

Reply:
Joined and dm’d

Reply:
You should be able to use this code to connect to a vault server that requires client cert auth (mTLS) and the vault server's cert is issued by an external CA (it will also work if the server's cert is issued by the Puppet CA).

```
require 'puppet'

def add_external_trust_store(ssl_context)
  path = Puppet[:ssl_trust_store]
  if path
    stat = Puppet::FileSystem.stat(path)
    if stat
      if stat.ftype == 'file'
        # don't add empty files as ruby/openssl will raise
        if stat.size > 0
          begin
            ssl_context.store.add_file(path)
          rescue => e
            Puppet.err(_("Failed to add '%{path}' as a trusted CA file: %{detail}" % { path: path, detail: e.message }, e))
          end
        end
      else
        Puppet.warning(_("The 'ssl_trust_store' setting does not refer to a file and will be ignored: '%{path}'" % { path: path }))
      end
    end
  end
end

Puppet.initialize_settings
Puppet[:ssl_trust_store] = " /etc/ssl/certs/ca-certificates.crt"

ssl_provider = Puppet::SSL::SSLProvider.new
ssl_context = ssl_provider.load_context(revocation: false)
ssl_context.store.set_default_paths
add_external_trust_store(ssl_context)

client = Puppet.runtime[:http]
client.get(URI("https://127.0.0.1:8888"), options: {ssl_context: ssl_context})
```

This should work for all puppet versions from 6.16.0 and up. In the future, once https://tickets.puppetlabs.com/browse/PUP-11522 is fixed, it will be possible to just call:

```ruby
client.get(URI("https://127.0.0.1:8888"), options: {include_system_store: true})
```

Reply:
in my setup when I uploaded CA to Vault, I appended Puppet CA with Vault server CA.

Reply:
> I appended Puppet CA with Vault server CA.

@vamsinm Could you expand on that? Did you add the vault server CA to the CA bundle that agents download from puppetserver? Or did you manually add the vault server CA to `/opt/puppetlabs/puppet/ssl/cert.pem` on each agent?

Could you save the above code to `sslclient.rb` , edit the vault URL & `ssl_trust_store` setting, try running it using puppet's ruby and verify it can connect to your vault server?

```
$ /opt/puppetlabs/puppet/bin/ruby sslclient.rb
```

Reply:
our puppet CA and CA for the Certificate that Vault server uses are different. 
i take puppet CA and append enterprise root CA to it. Use the appeneded CA file when create Vault Certificate role. using command.
```shell
vault write auth/cert/certs/puppetserver \
    display_name=puppet \
    policies=vault-puppet-policy \
    certificate=@puppet_ca_updated.pem \
    ttl=3600
```
nothing changed on puppet server or puppet agents.